### ISSUE TITLE
Dynamic parallelism does not work in cmsRun

### ISSUE THREAD

**Issue by VinInn** [cms-patatrack/cmssw#402]
just add
```
cat RecoPixelVertexing/PixelTriplets/plugins/crash.cu
__global__
void crash() {
 cudaDeviceSynchronize();
}
```

and cmsRun will crash as soon as it tries to run a kernel from that plugin....

with error
/home/innocent/ORI/CMSSW_11_0_0_pre11_Patatrack/src/RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernels.cu, line 175:
cudaCheck(cudaGetLastError());
cudaErrorInvalidDeviceFunction: invalid device function

adding -rdc true to nvcc does not help

in tests in CMSSW (no link to plugins) work fine

**Comment by VinInn** [comment]
according to "mps" doc

Dynamic parallelism is not supported. CUDA module load will fail if the module uses dynamic parallelism features.


so either mps or Dynamic parallelism

**Comment by fwyzard** [comment]
@VinInn could you see if we have the same issue also in CMSSW_10_6_0_Patatrack ?

**Comment by VinInn** [comment]
it seems working...
I was sure I tested as soon as I saw the issue with mps (the error is the same!)
Anyhow, no crash
and just to make sure
```
[innocent@patatrack02 src]$ nm -C $CMSSW_BASE/lib/slc7_amd64_gcc820/pluginRecoPixelVertexingPixelTripletsPlugins.so | grep  crash
000000000003b1e0 T __device_stub__Z5crashv()
000000000003b260 T crash()
00000000003cde67 r def_module_id_str_51_tmpxft_0000590e_00000000_8_crash_compute_70_cpp1_ii_b9424260
0000000000ce6190 b __cudaRegisterLinkedBinary_51_tmpxft_0000590e_00000000_8_crash_compute_70_cpp1_ii_b9424260::__p
0000000000074d20 T __cudaRegisterLinkedBinary_51_tmpxft_0000590e_00000000_8_crash_compute_70_cpp1_ii_b9424260
0000000000ce5f98 D __fatbinwrap_51_tmpxft_0000590e_00000000_8_crash_compute_70_cpp1_ii_b9424260

[innocent@patatrack02 src]$ cat $CMSSW_BASE/src/RecoPixelVertexing/PixelTriplets/plugins/crash.cu
__global__
void crash() {
 cudaDeviceSynchronize();
}
```

**Comment by VinInn** [comment]
Still, under the assumption we find a solution, we need to decide:
either mps or Dynamic parallelism

and We need to ask Nvidia if they plan to support Dynamic parallelism under mps in the short term

**Comment by fwyzard** [comment]
OK, so
  - with 10.6.0 (CUDA 10.1.105), both MPS and dynamic parallelism are working
  - with 10.6.3 and later (CUDA 10.1.243), neither is working

Tomorrow I'll try to see if something changed in the way NVCC builds our code, and send a report to Andreas Hehn with NVIDIA ...

**Comment by makortel** [comment]
> Still, under the assumption we find a solution, we need to decide:
> either mps or Dynamic parallelism

If we could foresee that I/O could be made to scale to full node (in both HLT and offline), I would prefer dynamic parallelism. But I'm not sure how realistic that is...

**Comment by VinInn** [comment]
what about the nvcc option
```
--relocatable-device-code {true|false} (-rdc)
```
is true the default?

**Comment by fwyzard** [comment]
@VinInn is this fixed with CUDA 11.0 ?

**Comment by VinInn** [comment]
No, not fixed
